In [120]:
opcodes={
    'B':{
        'name':'B',
        'range':[0b1,63-0],
        'codes':{
            0o1:{'mn':'&'}
        }
    },
    'SOP':{
        'name':'SOP',
        'range':[0b111,63-3],
        'codes':{
            0o0:{
                'name':'SOP1',
                'range':[0b1111,63-9],
                'disable':['SPS'],
                'codes':{
                    0o0:{'disable':['SPD']},
                    0o1:{'mn':'WRTEXP'},
                    0o2:{'mn':'WRTMH'},
                    0o3:{'mn':'WRTML'},
                    0o10:{'mn':'CLR'},
                    0o11:{'mn':'INC'},
                    0o12:{'mn':'DEC'},
                    0o13:{'mn':'COM'},
                    0o14:{'mn':'LOSPNL'},
                    0o15:{'mn':'LDSPE'},
                    0o16:{'mn':'LDSPI'},
                    0o17:{'mn':'LDSPT'}
                }
            },
            0o1:{
                'name':'SPEC',
                'range':[0b1111,63-9],
                'disable':['SPS','SPD'],
                'codes':{
                    0o0:{
                        'name':'STEST',
                        'range':[0b1111,63-13],
                        'codes':{
                            0o0:{'mn':'BFLT'},
                            0o1:{'mn':'BLT'},
                            0o2:{'mn':'BNC'},
                            0o3:{'mn':'BZC'},
                            0o4:{'mn':'BDBN'},
                            0o5:{'mn':'BDBZ'},
                            0o6:{'mn':'BIFN'},
                            0o7:{'mn':'BIFZ'},
                            0o14:{'mn':'BFL0'},
                            0o15:{'mn':'BFL1'},
                            0o16:{'mn':'BFL2'},
                            0o17:{'mn':'BFL3'},
                        }
                    },
                    0o1:{
                        'name':'HOSTPNL',
                        'range':[0b1111,63-13],
                        'codes':{
                            0o0:{'mn':'PNLLIT'},
                            0o1:{'mn':'DBELIT'},
                            0o2:{'mn':'DBHLIT'},
                            0o3:{'mn':'DBLLIT'},
                            0o10:{'mn':'SWDB'},
                            0o11:{'mn':'SWDBE'},
                            0o12:{'mn':'SWDBH'},
                            0o13:{'mn':'SWDBL'}
                        }
                    },
                    0o2:{'mn':'SPMDA'},
                    0o10:{
                        'name':'SETPSA',
                        'range':[0b1111,63-13],
                        'codes':{
                            0o0:{'mn':'JMPA','value':True},
                            0o1:{'mn':'JSRA','value':True},
                            0o2:{'mn':'JMP','value':True},
                            0o3:{'mn':'JSR','value':True},
                            0o4:{'mn':'JMPT'},
                            0o5:{'mn':'JSRT'},
                            0o6:{'mn':'JMPP'},
                            0o7:{'mn':'JSRP'}
                        }
                    },
                    0o11:{
                        'name':'PSEVEN',
                        'range':[0b1111,63-13],
                        'codes':{
                            0o0:{'mn':'RPS0A','value':True},
                            0o1:{'mn':'RPS2A','value':True},
                            0o2:{'mn':'RPS0','value':True},
                            0o3:{'mn':'RPS2','value':True},
                            0o4:{'mn':'RPS0T'},
                            0o5:{'mn':'RPS2T'},
                            0o10:{'mn':'WPS0A','value':True},
                            0o11:{'mn':'WPS2A','value':True},
                            0o12:{'mn':'WPS0','value':True},
                            0o13:{'mn':'WPS2','value':True},
                            0o14:{'mn':'WPS0T'},
                            0o15:{'mn':'WPS2T'}
                        }
                    },
                    0o12:{
                        'name':'PSODD',
                        'range':[0b1111,63-13],
                        'codes':{
                            0o0:{'mn':'RPS1A','value':True},
                            0o1:{'mn':'RPS3A','value':True},
                            0o2:{'mn':'RPS1','value':True},
                            0o3:{'mn':'RPS3','value':True},
                            0o4:{'mn':'RPS1T'},
                            0o5:{'mn':'RPS3T'},
                            0o10:{'mn':'WPS1A','value':True},
                            0o11:{'mn':'WPS3A','value':True},
                            0o12:{'mn':'WPS1','value':True},
                            0o13:{'mn':'WPS3','value':True},
                            0o14:{'mn':'WPS1T'},
                            0o15:{'mn':'WPS3T'}
                        }
                    },
                    0o13:{
                        'name':'PS',
                        'range':[0b1111,63-13],
                        'codes':{
                            0o0:{'mn':'RPSLA','value':True},
                            0o1:{'mn':'RPSFA','value':True},
                            0o2:{'mn':'RPSL','value':True},
                            0o3:{'mn':'RPSF','value':True},
                            0o4:{'mn':'RPSLT'},
                            0o5:{'mn':'RPSFT'},
                            0o6:{'mn':'RPSLP'},
                            0o7:{'mn':'RPSFP'},
                            0o10:{'mn':'LPSLA','value':True},
                            0o11:{'mn':'LPSFA','value':True},
                            0o12:{'mn':'LPSL','value':True},
                            0o13:{'mn':'LPSF','value':True},
                            0o14:{'mn':'LPSLT'},
                            0o15:{'mn':'LPSFT'},
                            0o16:{'mn':'LPSLP'},
                            0o17:{'mn':'LPSFP'}
                        }
                    },
                    0o14:{
                        'name':'SETEXIT',
                        'range':[0b1111,63-13],
                        'codes':{
                            0o1:{'mn':'SETEXA','value':True},
                            0o3:{'mn':'SETEX','value':True},
                            0o5:{'mn':'SETEXT'},
                            0o7:{'mn':'SETEXP'}
                        }
                    }
                }
            },
            0o2:{'mn':'ADD'},
            0o3:{'mn':'SUB'},
            0o4:{'mn':'MOV'},
            0o5:{'mn':'AND'},
            0o6:{'mn':'OR'},
            0o7:{'mn':'EQV'},
        }
    },
    'SH':{
        'name':'SH',
        'range':[0b11,63-5],
        'codes':{
            0o1:{'mn':'L'},
            0o2:{'mn':'RR'},
            0o3:{'mn':'R'}
        }
    },
    'SPS':{
        'name':'SPS',
        'range':[0b1111,63-9],
        'data':True
    },
    'SPD':{
        'name':'SPD',
        'range':[0b1111,63-13],
        'data':True
    },
    'FADD':{
        'name':'FADD',
        'range':[0b111,63-16],
        'codes':{
            0o0:{
                'name':'FADD1',
                'range':[0b111,63-19],
                'disable':['A1'],
                'codes':{
                    0o0:{'disable':['A2']},
                    0o1:{'mn':'FIX'},
                    0o2:{'mn':'FIXT'},
                    0o3:{'mn':'FSCLT'},
                    0o4:{'mn':'FSM2C'},
                    0o5:{'mn':'F2CSM'},
                    0o6:{'mn':'FSCALE'},
                    0o7:{'mn':'FABS'}
                }
            },
            0o1:{'mn':'FSUBR'},
            0o2:{'mn':'FSUB'},
            0o3:{'mn':'FADD'},
            0o4:{'mn':'FEQV'},
            0o5:{'mn':'FAND'},
            0o6:{'mn':'FOR'},
            0o7:{
                'name':'IO',
                'range':[0b111,63-19],
                'disable':['A1','A2'],
                'codes':{
                    0o0:{
                        'name':'LDREG',
                        'range':[0b111,63-22],
                        'codes':{
                            0o1:{'mn':'LDSPD'},
                            0o2:{'mn':'LDMA'},
                            0o3:{'mn':'LDTMA'},
                            0o4:{'mn':'LDDPA'},
                            0o5:{'mn':'LDDPA'},
                            0o6:{'mn':'LDSP'},
                            0o7:{'mn':'LDDA'}
                        }
                    },
                    0o1:{
                        'name':'RDREG',
                        'range':[0b111,63-22],
                        'codes':{
                            0o0:{'mn':'RPSA'},
                            0o1:{'mn':'RSPD'},
                            0o2:{'mn':'RMA'},
                            0o3:{'mn':'RTMA'},
                            0o4:{'mn':'RSP'},
                            0o5:{'mn':'RDPA'},
                            0o6:{'mn':'RSP'},
                            0o7:{'mn':'RDA'}
                        }
                    },
                    0o2:{'mn':'SPMDAV'},
                    0o3:{'mn':'REXIT'},
                    0o4:{
                        'name':'INOUT',
                        'range':[0b111,63-22],
                        'codes':{
                            0o0:{'mn':'OUT'},
                            0o1:{'mn':'SPNOUT'},
                            0o2:{'mn':'OUTDA'},
                            0o3:{'mn':'SPOTDA'},
                            0o4:{'mn':'IN'},
                            0o5:{'mn':'SPININ'},
                            0o6:{'mn':'INDA'},
                            0o7:{'mn':'SPINDA'}
                        }
                    },
                    0o5:{
                        'name':'SENCE',
                        'range':[0b111,63-22],
                        'codes':{
                            0o0:{'mn':'SNSA'},
                            0o1:{'mn':'SPINA'},
                            0o2:{'mn':'SNSADA'},
                            0o3:{'mn':'SPNADA'},
                            0o4:{'mn':'SNSB'},
                            0o5:{'mn':'SPINB'},
                            0o6:{'mn':'SNSBDA'},
                            0o7:{'mn':'SPNBDA'}
                        }
                    },
                    0o6:{
                        'name':'FLAG',
                        'range':[0b111,63-22],
                        'codes':{
                            0o0:{'mn':'SFL0'},
                            0o1:{'mn':'SFL1'},
                            0o2:{'mn':'SFL2'},
                            0o3:{'mn':'SFL3'},
                            0o4:{'mn':'CFL0'},
                            0o5:{'mn':'CFL1'},
                            0o6:{'mn':'CFL2'},
                            0o7:{'mn':'CFL3'}
                        }
                    },
                    0o7:{
                        'name':'CONTROL',
                        'range':[0b111,63-22],
                        'codes':{
                            0o0:{'mn':'HALT'},
                            0o1:{'mn':'IORST'},
                            0o2:{'mn':'INTEN'},
                            0o3:{'mn':'INTA'},
                            0o4:{'mn':'REFR'},
                            0o5:{'mn':'WRTEX'},
                            0o6:{'mn':'WRTMAN'}
                        }
                    }
                }
            }
        }
    },
    'A1':{
        'name':'A1',
        'range':[0b111,63-19],
        'codes':{
            0o0:{'mn':'NC'},
            0o1:{'mn':'FM'},
            0o2:{'mn':'DPX(IDX)','IDX':'XR'},
            0o3:{'mn':'DPY(IDX)','IDX':'YR'},
            0o4:{'mn':'TM'},
            0o5:{'mn':'ZERO'},
        }
    },
    'A2':{
        'name':'A2',
        'range':[0b111,63-22],
        'codes':{
            0o0:{'mn':'NC'},
            0o1:{'mn':'FM'},
            0o2:{'mn':'DPX(IDX)','IDX':'XR'},
            0o3:{'mn':'DPY(IDX)','IDX':'YR'},
            0o4:{'mn':'MD'},
            0o5:{'mn':'ZERO'},
            0o6:{'mn':'MDPX(IDX)','IDX':'XR'},
            0o7:{'mn':'EDPX(IDX)','IDX':'XR'},
        }
    },
    'COND':{
        'name':'COND',
        'range':[0b1111,63-26],
        'codes':{
            0o0:{'disable':['DISP']},
            0o1:{'mn':'#'},
            0o2:{'mn':'BR'},
            0o3:{'mn':'BINTRQ'},
            0o4:{'mn':'BION'},
            0o5:{'mn':'BIOZ'},
            0o6:{'mn':'BFPE'},
            0o7:{'mn':'RETURN'},
            0o11:{'mn':'BFEQ'},
            0o12:{'mn':'BFNE'},
            0o13:{'mn':'BFGT'},
            0o14:{'mn':'BEQ'},
            0o15:{'mn':'BNE'},
            0o16:{'mn':'BGE'},
            0o17:{'mn':'BGT'}
        }
    },
    'DISP':{
        'name':'DISP',
        'range':[0b11111,63-31],
        'data':True
    },
    'DPX':{
        'name':'DPX',
        'range':[0b11,63-33],
        'codes':{
            0o1:{'mn':'DPX(IDX)<DB','IDX':'XW'},
            0o2:{'mn':'DPX(IDX)<FA','IDX':'XW'},
            0o3:{'mn':'DPX(IDX)<FM','IDX':'XW'}
        }
    },
    'DPY':{
        'name':'DPX',
        'range':[0b11,63-35],
        'codes':{
            0o1:{'mn':'DPY(IDX)<DB','IDX':'YW'},
            0o2:{'mn':'DPY(IDX)<FA','IDX':'YW'},
            0o3:{'mn':'DPY(IDX)<FM','IDX':'YW'}
        }
    },
    'DPBS':{
        'name':'DPBS',
        'range':[0b111,63-38],
        'codes':{
            0o0:{'mn':'DB=ZERO'},
            0o1:{'mn':'DB=INSB'},
            0o2:{'mn':'DB=VALUE','value':True},
            0o3:{'mn':'DB=DPX(IDX)','IDX':'XR'},
            0o4:{'mn':'DB=DPY(IDX)','IDX':'YR'},
            0o5:{'mn':'DB=MD'},
            0o6:{'mn':'DB=SPFN'},
            0o7:{'mn':'DB=TM'}
        }
    },
    'XR':{
        'name':'XR',
        'range':[0b111,63-41]
    },
    'YR':{
        'name':'YR',
        'range':[0b111,63-44]
    },
    'XW':{
        'name':'XW',
        'range':[0b111,63-47]
    },
    'YW':{
        'name':'YW',
        'range':[0b111,63-50]
    },
    'FM':{
        'name':'FM',
        'range':[0b1,63-51],
        'codes':{
            0o0:{'disable':['M1','M2']},
            0o1:{'mn':'FMUL'}
        }
    },
    'M1':{
        'name':'M1',
        'range':[0b11,63-53],
        'codes':{
            0o0:{'mn':'FM'},
            0o1:{'mn':'DPX(IDX)','IDX':'XR'},
            0o2:{'mn':'DPY(IDX)','IDX':'YR'},
            0o3:{'mn':'TM'},
        }
    },
    'M2':{
        'name':'M2',
        'range':[0b11,63-55],
        'codes':{
            0o0:{'mn':'FA'},
            0o1:{'mn':'DPX(IDX)','IDX':'XR'},
            0o2:{'mn':'DPY(IDX)','IDX':'YR'},
            0o3:{'mn':'MD'},
        }
    },
    'MI':{
        'name':'MI',
        'range':[0b11,63-57],
        'codes':{
            0o1:{'mn':'MI<FA'},
            0o2:{'mn':'MI<FM'},
            0o3:{'mn':'MI<DB'},
        }
    },
    'MA':{
        'name':'MA',
        'range':[0b11,63-59],
        'codes':{
            0o1:{'mn':'INCMA'},
            0o2:{'mn':'DECMA'},
            0o3:{'mn':'SETMA'},
        }
    },
    'DPA':{
        'name':'DPA',
        'range':[0b11,63-61],
        'codes':{
            0o1:{'mn':'INCDPA'},
            0o2:{'mn':'DECDPA'},
            0o3:{'mn':'SETDPA'},
        }
    },
    'TMA':{
        'name':'TMA',
        'range':[0b11,63-63],
        'codes':{
            0o1:{'mn':'INCTMA'},
            0o2:{'mn':'DECTMA'},
            0o3:{'mn':'SETTMA'},
        }
    }
}

In [100]:
def getcode(val):
    global opcodes
    if opcodes is not None:
        op=opcodes.get(val.pop(0),{})
        while len(val)>0:
            if 'codes' in op:
                op=op['codes']
            op=op.get(val.pop(0),{})
        if 'mn' in op:
            return f"({op['mn']})"
        elif 'name' in op:
            return f"({op['name']})"
    return ''

In [101]:
opchar='''
0001000
0000000
0000000
0000000
'''

In [102]:
lines=[line.strip() for line in opchar.splitlines() if line]
op0,op1,op2,op3=[int(op,8)&0xFFFF for op in lines[0:4]]
opcat=(op0<<48)|(op1<<32)|(op2<<16)|(op3)
printall=False
usevalue=False
print(oct(opcat),oct(op0),oct(op1),oct(op2),oct(op3))
B=(opcat>>(63-0))&0b1
SOP=(opcat>>(63-3))&0b111
SH=(opcat>>(63-5))&0b11
SPS=(opcat>>(63-9))&0b1111
SPD=(opcat>>(63-13))&0b1111
SOP1=SPS
SPEC=(opcat>>(63-13))&0b11111111
SPEC0=SPEC>>4
SPEC1=SPEC&0xF
usevalue|=(
       ((SPEC0==0o10) and SPEC1<0o4)                                    #SETPSA
    or ((SPEC0==0o11) and (SPEC1<0o4 or (SPEC1>=0o10 and SPEC1<0o14)))  #PSEVEN
    or ((SPEC0==0o12) and (SPEC1<0o4 or (SPEC1>=0o10 and SPEC1<0o14)))  #PSODD
    or ((SPEC0==0o13) and (SPEC1<0o4 or (SPEC1>=0o10 and SPEC1<0o14)))  #PS
    or ((SPEC0==0o14) and SPEC1 in(0o1,0o3))                            #SETEXIT
)
if printall:
    print(f'S-PAD: B:{B:o} SOP:{SOP:o} SH:{SH:o} SPS:{SPS:o} SPD:{SPD:o} SOP1:{SOP1:o} SPEC:{SPEC:o}')
elif SOP==0o0:
    print(f'S-PAD: B:{B:o} SOP:{SOP:o}{getcode(["SOP",SOP])} SH:{SH:o}{getcode(["SH",SH])} SOP1:{SOP1:o}{getcode(["SOP",SOP,SOP1])} SPD:{SPD:o}')
elif SOP==0o1:
    print(f'S-PAD: B:{B:o} SOP:{SOP:o}{getcode(["SOP",SOP])} SH:{SH:o}{getcode(["SH",SH])} SPEC0:{SPEC0:o}{getcode(["SOP",SOP,SPEC0])} SPEC1{SPEC1:o}{getcode(["SOP",SOP,SPEC0,SPEC1])}')
else:
    print(f'S-PAD: B:{B:o} SOP:{SOP:o}{getcode(["SOP",SOP])} SH:{SH:o}{getcode(["SH",SH])} SPS:{SPS:o} SPD:{SPD:o}')
FADD=(opcat>>(63-16))&0b111
A1=(opcat>>(63-19))&0b111
A2=(opcat>>(63-22))&0b111
FADD1=A1
IO=(opcat>>(63-22))&0b111111
IO0=IO>>3
IO1=IO&0b111
if printall:
    print(f'ADDER: FADD:{FADD:o} A1:{A1:o} A2:{A2:o} FADD1:{FADD1:o} IO:{IO:o}')
elif FADD==0o0:
    print(f'ADDER: FADD:{FADD:o}{getcode(["FADD",FADD])} FADD1:{FADD1:o}{getcode(["A2",FADD,FADD1])} A2:{A2:o}{getcode(["A2",A2])}')
elif FADD==0o7:
    print(f'ADDER: FADD:{FADD:o}{getcode(["FADD",FADD])} IO0:{IO0:o}{getcode(["FADD",FADD,IO0])} IO1:{IO1:o}{getcode(["FADD",FADD,IO0,IO1])}')
else:
    print(f'ADDER: FADD:{FADD:o}{getcode(["FADD",FADD])} A1:{A1:o}{getcode(["A1",A1])} A2:{A2:o}{getcode(["A2",A2])}')
COND=(opcat>>(63-26))&0b1111
DISP=(opcat>>(63-31))&0b1111
print(f'BRANCH: COND:{COND:o}{getcode(["COND",COND])} DISP:{DISP:o}')
DPX=(opcat>>(63-33))&0b11
DPY=(opcat>>(63-35))&0b11
DPBS=(opcat>>(63-38))&0b111
usevalue|=(DPBS==0o2)
XR=(opcat>>(63-41))&0b111
YR=(opcat>>(63-44))&0b111
XW=(opcat>>(63-47))&0b111
YW=(opcat>>(63-50))&0b111
if printall or not usevalue:
    print(f'DATA PAD: DPX:{DPX:o}{getcode(["DPX",DPX])} DPY:{DPY:o}{getcode(["DPY",DPY])} DPBS:{DPBS:o}\
{getcode(["DPBS",DPBS])} XR:{XR:o} YR:{YR:o} XW:{XW:o} YW:{YW:o}')
else:
    print(f'DATA PAD: DPX:{DPX:o}{getcode(["DPX",DPX])} DPY:{DPY:o}{getcode(["DPY",DPY])} DPBS:{DPBS:o}\
{getcode(["DPBS",DPBS])} XR:{XR:o} YR:{YR:o} XW YW:{XW:o}')
FM=(opcat>>(63-51))&0b1
M1=(opcat>>(63-53))&0b11
M2=(opcat>>(63-55))&0b11
if printall or not usevalue:
    print(f'MULTIPLY: FM:{FM:o}{getcode(["FM",FM])} M1:{M1:o}{getcode(["M1",M1])} M2:{M2:o}{getcode(["M2",M2])}')
MI=(opcat>>(63-57))&0b11
MA=(opcat>>(63-59))&0b11
DPA=(opcat>>(63-61))&0b11
TMA=(opcat>>(63-63))&0b11
if printall or not usevalue:  
    print(f'MEMORY: MI:{MI:o}{getcode(["MI",MI])} MA:{MA:o}{getcode(["MA",MA])} \
DPA:{DPA:o}{getcode(["DPA",DPA])} TMA:{TMA:o}{getcode(["TMA",TMA])}')
VALUE=(opcat>>(63-63))&0xFFFF
if printall or usevalue:
    print(f'VALUE:{VALUE:o}')

0o10000000000000000000 0o1000 0o0 0o0 0o0
S-PAD: B:0 SOP:0(SOP1) SH:0 SOP1:10(CLR) SPD:0
ADDER: FADD:0(FADD1) FADD1:0 A2:0(NC)
BRANCH: COND:0 DISP:0
DATA PAD: DPX:0 DPY:0 DPBS:0(DB=ZERO) XR:0 YR:0 XW YW:0
VALUE:0


In [123]:
def parce_code(opcat):
    global opcodes
    inst=[]
    value=False
    disabled=set()
    for code in opcodes.values():
        op=code
        if code.get('name') in disabled:
            continue
        mask,offset=op.get('range')
        val=(opcat>>offset)&mask
        while 'codes' in op:
            mask,offset=op.get('range')
            val=(opcat>>offset)&mask
            #print(op.get('name'),63-offset,mask,val)
            disabled.update(op.get('disable',[]))
            codes=op.get('codes',{})
            if val in codes:
                op=codes.get(val)
            else:
                break
        value|='value' in op
        disabled.update(op.get('disable',[]))
        if offset<16 and value:
            break
        mn=op.get('mn')
        name=op.get('name','')
        if mn:
            if 'IDX' in op:
                mask,offset=opcodes.get(op.get('IDX')).get('range')
                val=(opcat>>offset)&mask
                mn=mn.replace('IDX',f'{val-4}')
            inst.append(mn)
        if 'data' in op:
            inst.append(f'0{val:o}')
        #print(name,mn,63-offset,mask,(opcat>>offset)&mask,value,disabled)
    #print(disabled)
    sinst=";".join(inst)
    if value:
        if 'VALUE' in sinst:
            sinst=sinst.replace('VALUE',f'0{op3:o}')
        else:
            sinst+=f' VAL: 0{op3:o}'
    return sinst


In [126]:
lines=[line.strip() for line in opchar.splitlines() if line]
op0,op1,op2,op3=[int(op,8)&0xFFFF for op in lines[0:4]]
opcat=(op0<<48)|(op1<<32)|(op2<<16)|(op3)
sinst=parce_code(opcat)
print(oct(op0),oct(op1),oct(op2),oct(op3))
print(sinst)

0o1000 0o0 0o0 0o0
CLR;00;DB=ZERO


In [35]:
lines_list='''
;No-Op
0000000
0000000
0000000
0000000
;LDTMA;DB=!THIRD ; !THIRD is 04430
0000003
0103000
0002000
0004430
;LDMA;DB=@077
0000003
0102000
0002000
0000077
;LDDPA;DB=ZERO
0000003
0104000
0000000
0000000
;DPX(0)<DB;DB=TM
0000000
0000000
0047444
0100000
;LDTMA;DB=!SIXTN ; !SIXTN is 4451
0000003
0103000
0002000
0004451
;No-Op
0000000
0000000
0000000
0000000
;DPY(0)<DB;DB=TM
0000000
0000000
0017444
0100000
;FADD DPX,DPY
0000001
0123000
0000444
0100000
;FADD
0000001
0100000
0000000
0000000
;MI<FA;INCMA
0000000
0000000
0000000
0000120
;No-Op
0000000
0000000
0000000
0000000
;No-Op
0000000
0000000
0000000
0000000
;Halt
0000003
0170000
0000000
0000000
'''
lines_list=[line.strip() for line in lines_list.splitlines() if line and not line.startswith(';')]
lines_list=[lines_list[i:i + 4] for i in range(0, len(lines_list), 4)]

In [125]:
for lines in lines_list:
    lines=[line.strip() for line in lines if line]
    op0,op1,op2,op3=[int(op,8)&0xFFFF for op in lines[0:4]]
    opcat=(op0<<48)|(op1<<32)|(op2<<16)|(op3)
    sinst=parce_code(opcat)
    #print(oct(op0),oct(op1),oct(op2),oct(op3))
    print(sinst)


NameError: name 'lines_list' is not defined

In [143]:
#To clear all bits:
#S-PAD GROUP
B=0o0;SOP=0o0;SH=0o0;SPS=0o0;SPD=0o0
SOP1=0o0
SPEC0=0o0;SPEC1=0o0
#ADDER GROUP
FADD=0o0;A1=0o0;A2=0o0
FADD1=0o0
IO0=0o0;IO1=0o0
#BRANCH GROUP
COND=0o0;DISP=0o0
#DATA PAD GROUP
DPX=0o0;DPY=0o0;DPBS=0o0;XR=0o0;YR=0o0;XW=0o0;YW=0o0
#MULTIPLY GROUP
FM=0o0;M1=0o0;M2=0o0
#MEMORY GROUP
MI=0o0;MA=0o0;DPA=0o0;TMA=0o0
#VALUE
VALUE=0o0

In [144]:
# DEC 0; DPX(0)<DB;DB=ZERO
SOP1=0o12
SPD=0
DPX=1
XW=4
DPBS=0

In [145]:
opcat=0
opcat|=(B&0b1)<<(63-0)
opcat|=(SOP&0b111)<<(63-3)
opcat|=(SH&0b11)<<(63-5)
SPS|=SOP1
opcat|=(SPS&0b1111)<<(63-9)
opcat|=(SPD&0b1111)<<(63-13)
SPEC=(SPEC0&0xF)<<4
SPEC|=SPEC1&0xF
opcat|=(SPEC&0b11111111)<<(63-13)
opcat|=(FADD&0b111)<<(63-16)
A1|=FADD1
opcat|=(A1&0b111)<<(63-19)
opcat|=(A2&0b111)<<(63-22)
IO=(IO0&0b111)<<3
IO|=(IO1&0b111)
opcat|=(IO&0b111111)<<(63-22)
opcat|=(COND&0b1111)<<(63-26)
opcat|=(DISP&0b1111)<<(63-31)
opcat|=(DPX&0b11)<<(63-33)
opcat|=(DPY&0b11)<<(63-35)
opcat|=(DPBS)&0b111<<(63-38)
opcat|=(XR&0b111)<<(63-41)
opcat|=(YR&0b111)<<(63-44)
opcat|=(XW&0b111)<<(63-47)
opcat|=(YW&0b111)<<(63-50)
opcat|=(FM&0b1)<<(63-51)
opcat|=(M1&0b11)<<(63-53)
opcat|=(M2&0b11)<<(63-55)
opcat|=(MI&0b11)<<(63-57)
opcat|=(MA&0b11)<<(63-59)
opcat|=(DPA&0b11)<<(63-61)
opcat|=(TMA&0b11)<<(63-63)
opcat|=(VALUE&0xFFFF)
op0,op1,op2,op3=(opcat>>48)&0xFFFF,(opcat>>32)&0xFFFF,(opcat>>16)&0xFFFF,(opcat)&0xFFFF
print(oct(op0),oct(op1),oct(op2),oct(op3))
print(parce_code(opcat))
print(f'{op0:07o}\n{op1:07o}\n{op2:07o}\n{op3:07o}\n')

0o1200 0o0 0o40004 0o0
DEC;00;DPX(0)<DB;DB=ZERO
0001200
0000000
0040004
0000000



CLR;00;DB=ZERO;SETMA;SETDPA
0001000
0000000
0000000
0000074
DEC;00;DPX(0)<DB;DB=ZERO
0001200
0000000
0040004
0000000
